africanus/gridding/wgridder/tests/test_wgridder.py

```diff
@@ -0,0 +1,403 @@
+                   nthreads=nthreads).compute()
+
+    # relative error should agree to within epsilon
+    vmax = np.maximum(np.abs(vis).max(), np.abs(vis_da).max())
+    assert_array_almost_equal(vis/vmax, vis_da/vmax,
+                              decimal=decimal)
+
+
+@pmp("nx", (30, 250))
+@pmp("ny", (128,))
+@pmp("fov", (5.0,))
+@pmp("nrow", (3333, 10000))
+@pmp("nchan", (1, 8))
+@pmp("nband", (1, 2))
+@pmp("precision", ('single', 'double'))
+@pmp("nthreads", (1, 4))
+@pmp("nchunks", (1, 3))
+def test_dask_residual(nx, ny, fov, nrow, nchan, nband,
+                       precision, nthreads, nchunks):
+    da = pytest.importorskip("dask.array")
+    from africanus.gridding.wgridder import residual as residual_np
+    from africanus.gridding.wgridder.dask import residual
+    np.random.seed(420)
+    if precision == 'single':
+        real_type = np.float32
+        complex_type = np.complex64
+        decimal = 4  # does not pass at 5
+    else:
+        real_type = np.float64
+        complex_type = np.complex128
+        decimal = 7
+    cell = fov*np.pi/180/nx
+    f0 = 1e9
+    freq = (f0 + np.arange(nchan)*(f0/nchan))
+    uvw = ((np.random.rand(nrow, 3)-0.5) /
+           (cell*freq[-1]/lightspeed))
+    vis = (np.random.rand(nrow, nchan)-0.5 + 1j *
+           (np.random.rand(nrow, nchan)-0.5)).astype(complex_type)
+    wgt = np.random.rand(nrow, nchan).astype(real_type)
+    step = np.maximum(1, nchan//nband)
+    if step:
+        freq_bin_idx = np.arange(0, nchan, step)
+        freq_mapping = np.append(freq_bin_idx, nchan)
+        freq_bin_counts = freq_mapping[1::] - freq_mapping[0:-1]
+    else:
+        freq_bin_idx = np.array([0], dtype=np.int8)
+        freq_bin_counts = np.array([1], dtype=np.int8)
+    nband = freq_bin_idx.size
+    image = np.random.randn(nband, nx, ny).astype(real_type)
+    residim_np = residual_np(uvw, freq, image, vis, freq_bin_idx,
+                             freq_bin_counts, cell, weights=wgt,
+                             nthreads=nthreads)
+
+    rows_per_task = int(np.ceil(nrow/nchunks))
+    row_chunks = (nchunks-1) * (rows_per_task,)
+    row_chunks += (nrow - np.sum(row_chunks),)
+    freq_da = da.from_array(freq, chunks=step)
+    uvw_da = da.from_array(uvw, chunks=(row_chunks, -1))
+    image_da = da.from_array(image, chunks=(1, nx, ny))
+    vis_da = da.from_array(vis, chunks=(row_chunks, step))
+    wgt_da = da.from_array(wgt, chunks=(row_chunks, step))
+    freq_bin_idx_da = da.from_array(freq_bin_idx, chunks=1)
+    freq_bin_counts_da = da.from_array(freq_bin_counts, chunks=1)
+
+    residim_da = residual(uvw_da, freq_da, image_da, vis_da,
+                          freq_bin_idx_da, freq_bin_counts_da,
+                          cell, weights=wgt_da, nthreads=nthreads).compute()
+
+    # should agree to within epsilon
+    rmax = np.maximum(np.abs(residim_np).max(), np.abs(residim_da).max())
+    assert_array_almost_equal(
+        residim_np/rmax, residim_da/rmax, decimal=decimal)
```

Comment:
These tests look very comprehensive.

Reply:
I went a bit overboard on the parametrisations. Still, the full test suite for the wgridder runs in about 50s on my laptop